compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirConventionFunctionCallChecker.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.KtNodeTypes
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getChild
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.references.FirErrorNamedReference
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeUnresolvedNameError
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+object FirConventionFunctionCallChecker : FirFunctionCallChecker() {
+    override fun check(expression: FirFunctionCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        val calleeReference = expression.calleeReference as? FirErrorNamedReference ?: return
+        val diagnostic = calleeReference.diagnostic as? ConeUnresolvedNameError ?: return
```

Comment:
Those unresolved named error that will be handled here should be filtered out at `ConeDiagnostic.toFirDiagnostics` as we do for unsafe call variants from inapplicable candidate error? Otherwise, we will see duplicate diagnostics, and I think we're seeing those currently.

Reply:
Thanks! Added filtering logic in `ErrorNodeDiagnosticCollectorComponent`

Reply:
Wouldn't it be better to do all of this in `coneDiagnosticToFirDiagnostic.kt` rather than in a separate checker? It is confusing now to have coupled logic in 2 spots (filtering and error reporting). Something like:

```kt
fun ConeDiagnostic.toFirDiagnostic(source: FirSourceElement): FirDiagnostic<FirSourceElement>? = when (this) {
  //...
  is ConeUnresolvedNameError -> mapUnresolvedNameError(diagnostic, source)
  //...
}

private fun mapUnresolvedNameError(diagnostic: ConeUnresolvedNameError, source: FirSourceElement): FirDiagnostic<*> {
  if (source.kind == FirFakeSourceElementKind.ArrayAccessNameReference) {
    return when (diagnostic.name) {
      OperatorNameConventions.GET -> FirErrors.NO_GET_METHOD.on(source)
      OperatorNameConventions.SET -> FirErrors.NO_SET_METHOD.on(source)
    }
  }
  return FirErrors.UNRESOLVED_REFERENCE.on(source, diagnostic.name.asString())
}
```

Reply:
I agree it's confusing to have these in two locations. But the trouble is that in `coneDiagnosticToFirDiagnostic.kt` there is no way to get the exact source element to report. The way I see the situation is that, `coneDiagnosticToFirDiagnostic.kt` is simply trying to do too much ATM. Maybe it's OK to let it work this way as long as the particular diagnostics it reports are straightforward to report (easy to locate the source w/o relying on position strategy). But if a diagnostic is not straightforward to report (for example setter array access can be more than one level below the passed `source`), then it makes sense to move the logic to a devoted checker that has more contexts.

Reply:
Thanks for clarifying that the source isn't the same. IMO I don't think it's an issue to have to find the right source to report on in `coneDiagnosticToFirDiagnostic.kt`. We do that there for `UNSAFE_CALL` variants. `NO_(GET|SET)_METHOD` is a lot less complicated than what `FirDestructuringDeclarationChecker` does. But I'll defer to your judgment in this case...